### ISSUE TITLE
yubikey HSM driver not loading on OSX 10.11.6 (15G1004)

### ISSUE THREAD

**Issue by trinitronx** [docker/notary#1006]
Hello,  I'm trying to test out notary with the Yubikey 4 support, and having some troubles getting it to work.

I'm using `notary version`:

```
notary
 Version:    0.4.2
 Git commit: c8aa8cf
```

When running `notary init`, it logs an error loading `/usr/local/lib/libykcs11.dylib`:

```
notary init  example.com/collection -D -s http://notary-server:8080
DEBU[0000] Using the following trust directory: /Users/exampleuser/.notary
DEBU[0000] Trusting 1 certs
ERRO[0000] could not reach http://notary-server:8080: Get http://notary-server:8080/v2/: dial tcp 192.168.64.2:8080: getsockopt: connection refused
INFO[0000] continuing in offline mode
DEBU[0000] No yubikey found, using alternative key storage: found library /usr/local/lib/libykcs11.dylib, but initialize error pkcs11: 0x6: CKR_FUNCTION_FAILED
DEBU[0000] No yubikey found, using alternative key storage: found library /usr/local/lib/libykcs11.dylib, but initialize error pkcs11: 0x6: CKR_FUNCTION_FAILED
No root keys found. Generating a new root key...
DEBU[0000] generated ECDSA key with keyID: ad74b12de90f93ed5acad8daded968e5f4f95509906728cfb24bbc3b8b01c156
DEBU[0000] generated new ecdsa key for role: root and keyID: ad74b12de90f93ed5acad8daded968e5f4f95509906728cfb24bbc3b8b01c156
DEBU[0000] No yubikey found, using alternative key storage: found library /usr/local/lib/libykcs11.dylib, but initialize error pkcs11: 0x6: CKR_FUNCTION_FAILED
You are about to create a new root signing key passphrase. This passphrase
will be used to protect the most sensitive key in your signing system. Please
choose a long, complex passphrase and be careful to keep the password and the
key file itself secure and backed up. It is highly recommended that you use a
password manager to generate the passphrase and keep it safe. There will be no
way to recover this key. You can find the key in your config directory.
Enter passphrase for new root key with ID ad74b12:
```

**Comment by trinitronx** [comment]
I realize this is quite new and some features are probably beta still... however I'm having trouble understanding the process due to confusing or little documentation on using the Yubikey 4 with `notary`.

I've already gotten my Yubikey set up for use with GPG keys, and it works fine for signing via `gpg`, and shows up with `gpg --card-status` through `scdaemon` + `gpg-agent`.  However, it's unclear to me why `notary` is trying to generate a new root key when the Yubikey should already contain keys that I do not want to change or replace.  Is there a way to specify that I'd rather have notary use a delegate cert signed by my root Yubikey signing key (already in the Yubikey 4)?

**Comment by riyazdf** [comment]
Hi @trinitronx, sorry to hear you're running into this!

Notary seems to be finding the `libykcs11.dylib`, but it's having a [problem initializing](https://www.cryptsoft.com/pkcs11doc/v200/pkcs11__all_8h.html#aCKR_FUNCTION_FAILED) a pkcs11 session with it
(from this debug line: `DEBU[0000] No yubikey found, using alternative key storage: found library /usr/local/lib/libykcs11.dylib, but initialize error pkcs11: 0x6: CKR_FUNCTION_FAILED`)
And since it's failing to initialize a pkcs11 session with that library, it's not even getting around to using your yubikey and thusly falls back to creating an encrypted root key on disk.
@endophage @cyli please feel free to chime in if this is missing some detail or needs to be corrected.

How did you initially install `libykcs11.dylib`?  I used `brew install yubico-piv-tool` for a OSX install.  And did you use the same library for your `gpg` signing?

As for what kind of keys can be used: Notary currently only has support for root private keys (ECDSA/RSA) to be stored on a Yubikey, though we'd like to support [all key roles](https://github.com/docker/notary/issues/306) in the future.

**Comment by cyli** [comment]
@trinitronx re: overwriting keys:  Did you already have PIV keys on the yubikey (such as for OS X or windows code signing?), or just the GPG keys?  IIRC the PIV keys are completely separate from the GPG keys, and are stored in different slots on the card (I have my 3 gpg keys and a notary PIV key on mine, for instance).  I don't think we overwrite existing other PIV signing keys - we generate our own so that you don't end up using the same keys for everything - but we won't replace any we haven't created).

@trinitronx @riyazdf:  re: `initialize error pkcs11: 0x6: CKR_FUNCTION_FAILED`:  I remember hearing several months ago (before sierra) about how the some version of OS X broke the piv library.  I don't remember if that is still the case, or what version of OS X it was, however.  For very old versions of OS X I remember Yubico had a libccid driver that had to be installed as well - if the library was there and the driver is not (or there was some other issue where the library could not access the card) that could account for the initialization error.  Deferring to @endophage as I haven't played with the drivers in a long time.

**Comment by endophage** [comment]
The first thing to be aware of is that Yubikeys contain physically distinct internal hardware for GPG vs PIV/PKCS11 (the part Notary utilizes). Your GPG keys, and your PKCS11 keys are mutually exclusive sets. These features also operate over different protocols, so GPG working provides no kind of indicator as to whether the PKCS11 part of the Yubikey is functioning.

The first step would be to ensure your Yubikey is actually configured correctly. Yubikeys can be configured in multiple modes and it's possible that you are in a mode that does not support PKCS11. You can find the info by clicking the Apple icon in your menubar, selecting "About This Mac", and opening System Report. Under the USB section you should check that "CCID" is included in the Yubikey's device name. More specifically, the Product ID should be one of `0x0405`, `0x0406`, or `0x0407`.

![image](https://cloud.githubusercontent.com/assets/1302091/19373645/8d9f89f0-917a-11e6-952a-0db1597dea60.png)

If that's not the case, follow [these instructions](https://developers.yubico.com/libu2f-host/Mode_switch_YubiKey.html) to switch mode to something that includes CCID.

The next thing to check is that you have a sufficiently new version of udev. If you install yubico-piv-tool from brew, you should be fine. Try running `yubico-piv-tool -a status`. The output below is that command run with my Yubikey inserted, and subsequently again with the Yubikey removed, so you can see the difference in what's expected:

```
david$ yubico-piv-tool -a status
CHUID:	No data available
CCC:	No data available
PIN tries left:	3
david$ yubico-piv-tool -a status
Failed to connect to reader.
```

If none of the above raises anything obvious, we'll have to dig deeper, but it's typically one of those things.

**Comment by trinitronx** [comment]
@riyazdf: Yes I've installed `libykcs11.dylib` with the `yubico-piv-tool` with homebrew via `brew install yubico-piv-tool` as you recommend.  I did not previously have the `yubico-piv-tool` installed when I set up the key for doing `gpg` signing, so this library is not used for signing.  I believe `gpg-agent` starts `scdaemon` to ask the Yubikey to sign data.  I have followed [this guide][1] for initially setting up my Yubikey, however since I am using a Yubikey 4 nano, I can use 4096 bit keys instead of 2048 which was the previous limit for the Yubikey neo for which this guide was written.  I was previously using the stock version of [GPGTools for mac][2] (`v2.0.30`) installed via homebrew also: `brew cask install gpgtools`.  However, I have recently [upgraded to the newer `gpg` `v2.1.15`][3] for use with `gpg-agent` socket forwarding over ssh.  Both versions of GPG and included `scdaemon` work with the Yubikey 4.  The only issue I'm currently running into with `scdaemon` is that sometimes if left plugged in for a long time & the Macbook goes to sleep, `scdaemon` appears to get into a state where it cannot contact the Yubikey.  After killing `scdaemon`, and re-running `gpg2 --card-status`, it can then contact the Yubikey again.

As far as the `yubico-piv-tool` goes, it does detect my Yubikey 4 and prints out some info (I've scrubbed the CHUID with random data for privacy):

```
yubico-piv-tool -a status
CHUID:  68bfee79d359efdf74999f6dfb045692515e81ec5b16ff11828e3707d3d07c7d969c2eca8663a2cb946b47b11d0a9664d1bea90a79418cb7179cdd
CCC:  No data available
PIN tries left: 3
```

It's interesting that the `CCC` line says `No data available`.  I'm not sure what this portion of the output indicates, or how to determine if I'm missing the `libccid` driver as @cyli suggests.

It's surprising that the `PKCS11` PIV keys and GPG keys are in separate slots!  Amazing how much can fit on a little Yubikey 4 nano.  The `gpg2 --card-status` already shows 5 keys (1 primary + 4 subkeys), and `gpg2 --list-secret-keys` shows the subkeys are already of different types (ElGamal Encryption, Authentication, Signing, and RSA Encryption).

It appears that my Yubikey 4 does have the `Product ID` of `0x0407`, and contains `CCID` in the name:

[<img width="287" alt="screenshot 2016-10-14 11 06 38" src="https://cloud.githubusercontent.com/assets/122524/19396122/a64a6676-91fe-11e6-80fb-b7fc31f14b51.png">][4]


[1]: https://www.isi.edu/~calvin/yubikeyssh.htm
[2]: https://gpgtools.org/
[3]: https://www.rempe.us/blog/yubikey-gnupg-2-1-and-ssh/
[4]: https://www.dropbox.com/s/u5ritu18faxnlqf/Screenshot%202016-10-14%2011.06.38.png?dl=0

**Comment by endophage** [comment]
If the yubico-piv-tool is showing the key as being available, notary should be picking it up. I've just run a `notary init` with the same version of the client you're using. Could you also confirm you're using version 1.4.0 of the yubico-piv-tool: `yubico-piv-tool --version`. That would affect which version of libykcs11 you have.

Just to cover all the overt differences, the other thing I see is your Yubikey is 4.28. I have versions 4.26 and 4.31 available and they are both working as expected so I don't think that's an issue.

Unfortunately the pkcs11 interface is a bit of a blackbox when trying to debug remotely :-(

**Comment by endophage** [comment]
One other question, have you restarted since installing yubico-piv-tool? It has in the past needed to patch some system libraries and a restart was necessary to apply the patches.

**Comment by riyazdf** [comment]
Another question: is CCID mode enabled on your yubikey?  I remember this being configurable with `ykpersonalize -m6` or a similar mode flag.  Yubico also has some tools for helping install libccid and enabling this mode: https://developers.yubico.com/yubikey-neo-manager/Usage.html

**Comment by endophage** [comment]
@riyazdf CCID mode confirmed. The Yubikey is showing 0x0407 which is the "everything enabled" mode.

**Comment by trinitronx** [comment]
@endophage :   The version of `yubico-piv-tool` I'm using is:

```
$ yubico-piv-tool --version
yubico-piv-tool 1.4.0
```

Yubikey firmware version as reported via the `gpg-agent` is:

```
gpg-connect-agent --hex "scd apdu 00 f1 00 00" /bye
D[0000]  04 02 08 90 00                                     .....
OK
```

This lines up with the reported version from `lsusb` and the `Version` reported from `About this Mac -> System Report`: `4.28`


I did not reboot yesterday after installing `yubico-piv-tool` & trying out `notary`, however I can test it out again today after the reboot.

**Comment by trinitronx** [comment]
The `CCID` mode and `PIV` applet appears to be enabled and available as reported from the "YubiKey NEO Manager" program:

[![screenshot 2016-10-14 12 41 50](https://cloud.githubusercontent.com/assets/122524/19398829/b800a47c-920b-11e6-9eef-660fe4394be3.png)][1]

[![screenshot 2016-10-14 12 45 41](https://cloud.githubusercontent.com/assets/122524/19399038/b1cedb5e-920c-11e6-8fc4-548b027d511b.png)][2]


[1]: https://www.dropbox.com/s/oznel5vhr5dmfae/Screenshot%202016-10-14%2012.41.50.png?dl=0
[2]: https://www.dropbox.com/s/otgwjdt8y4t9v6f/Screenshot%202016-10-14%2012.45.41.png?dl=0

**Comment by endophage** [comment]
We'll wait on your reboot before looking into anything else.

**Comment by endophage** [comment]
We'll wait on your reboot before trying anything else.

**Comment by trinitronx** [comment]
@endophage:  Ok, so I tested again today after rebooting and it's still not able to access the Yubikey:

```
notary init -D  docker.io/trinitronx/build-tools
DEBU[0000] Using the following trust directory: /Users/exampleuser/.notary
DEBU[0000] Trusting 1 certs
DEBU[0000] No yubikey found, using alternative key storage: found library /usr/local/lib/libykcs11.dylib, but initialize error pkcs11: 0x6: CKR_FUNCTION_FAILED
DEBU[0000] No yubikey found, using alternative key storage: found library /usr/local/lib/libykcs11.dylib, but initialize error pkcs11: 0x6: CKR_FUNCTION_FAILED
No root keys found. Generating a new root key...
DEBU[0000] generated ECDSA key with keyID: 478f95c128b379b563e9e192a8fac8d438f32c5e3619f265bb1ca0d60f214a52
DEBU[0000] generated new ecdsa key for role: root and keyID: 478f95c128b379b563e9e192a8fac8d438f32c5e3619f265bb1ca0d60f214a52
DEBU[0000] No yubikey found, using alternative key storage: found library /usr/local/lib/libykcs11.dylib, but initialize error pkcs11: 0x6: CKR_FUNCTION_FAILED
You are about to create a new root signing key passphrase. This passphrase
will be used to protect the most sensitive key in your signing system. Please
choose a long, complex passphrase and be careful to keep the password and the
key file itself secure and backed up. It is highly recommended that you use a
password manager to generate the passphrase and keep it safe. There will be no
way to recover this key. You can find the key in your config directory.
Enter passphrase for new root key with ID 478f95c:
```

**Comment by endophage** [comment]
@a-dma sorry to bug you but we've run out of ideas on why @trinitronx isn't able to get his yubikey working with Notary. I thought you might be able to provide some more direction. It appears yubico-piv-tool, installed from brew, is of the same version I'm using (so I assume libykcs11 will be of the correct version), we're on the same version of Mac OS 10.11.6, the Yubikey is showing mode 0x0407 which means CCID is enabled, we've even tried a reboot (I couldn't remember 100% but thought that was necessary for the potential patches to take effect).

**Comment by trinitronx** [comment]
# Update:

So, on a hunch after reading [some bug reports][1] about problems with `scdaemon` Yubikey access I decided to test something out:

1. Verify that `gpg2 --card-status` is able to contact Yubikey & list keys
2. Kill all running `scdaemon` processes (which appear to get started by `gpg-agent` on-demand automatically)
3. Retry the `notary init -D` command

## Results:

```
notary init -D  docker.io/trinitronx/build-tools
DEBU[0000] Using the following trust directory: /Users/exampleuser/.notary
DEBU[0000] Trusting 1 certs
DEBU[0000] Initialized PKCS11 library /usr/local/lib/libykcs11.dylib and started HSM session
DEBU[0000] Failed to list key from the yubikey: No keys found in yubikey.
DEBU[0000] Initialized PKCS11 library /usr/local/lib/libykcs11.dylib and started HSM session
DEBU[0000] Failed to list key from the yubikey: No keys found in yubikey.
No root keys found. Generating a new root key...
DEBU[0000] generated ECDSA key with keyID: 2d39526716b1c3ef236892701bd6bc0c342a98246748d8a4e8c2de39e8de11ff
DEBU[0000] generated new ecdsa key for role: root and keyID: 2d39526716b1c3ef236892701bd6bc0c342a98246748d8a4e8c2de39e8de11ff
DEBU[0000] Initialized PKCS11 library /usr/local/lib/libykcs11.dylib and started HSM session
DEBU[0000] Attempting to store key using yubikey slot [2]
DEBU[0000] Attempting to add key to yubikey with ID: 2d39526716b1c3ef236892701bd6bc0c342a98246748d8a4e8c2de39e8de11ff
You are about to create a new root signing key passphrase. This passphrase
will be used to protect the most sensitive key in your signing system. Please
choose a long, complex passphrase and be careful to keep the password and the
key file itself secure and backed up. It is highly recommended that you use a
password manager to generate the passphrase and keep it safe. There will be no
way to recover this key. You can find the key in your config directory.
Enter passphrase for new root key with ID 2d39526:
```

## Observations:

It looks like the `scdaemon` process started by `gpg-agent` locks up the Yubikey for exclusive access, which prevents `notary` from accessing the key.  Killing `scdaemon` works for a brief period of time until the next `gpg` operation that needs to access the key & automatically starts up scdaemon again, blocking access to the key from anything else.

Additionally, there are bugs in the way `scdaemon` handles the Yubikey 4 on OSX, causing the key to become inaccessible even to the same `scdaemon` and `gpg-agent` processes that were able to access it before.  This issue begins to happen after some period of time in the [order of 30-mins to `1-2` (unknown?) hours][3] of time.

`scdaemon.log` file shows errors: `pcsc_connect failed: sharing violation (0x8010000b)`, `new_reader_slot: out of slots`.

Killing `scdaemon` temporarily fixed the access issue by GPG, then `gpg2 --card-status` worked.  Killing `scdaemon` again finally allowed `notary init -D` to work!

```
2016-10-17 09:49:03 scdaemon[2939] detected reader 'Yubico Yubikey 4 OTP+U2F+CCID'
2016-10-17 09:49:03 scdaemon[2939] pcsc_control failed: not transacted (0x80100016)
2016-10-17 09:49:03 scdaemon[2939] pcsc_vendor_specific_init: GET_FEATURE_REQUEST failed: 65547
2016-10-17 09:49:03 scdaemon[2939] updating reader 0 (0) status: 0x0000->0x0007 (0->1)
2016-10-17 09:49:03 scdaemon[2939] sending signal 31 to client 527
2016-10-17 09:54:03 scdaemon[2939] signatures created so far: 703
2016-10-17 09:54:03 scdaemon[2939] DBG: asking for PIN '||Please enter the PIN%0A[sigs done: 703]'
2016-10-17 10:32:18 scdaemon[2939] updating reader 0 (0) status: 0x0007->0x0000 (1->2)
2016-10-17 10:32:18 scdaemon[2939] DBG: Removal of a card: 0
2016-10-17 10:32:18 scdaemon[2939] DBG: application has been released
2016-10-17 10:32:18 scdaemon[2939] sending signal 31 to client 527
2016-10-17 10:53:50 scdaemon[2939] detected reader 'Yubico Yubikey 4 OTP+U2F+CCID'
2016-10-17 10:53:50 scdaemon[2939] pcsc_connect failed: sharing violation (0x8010000b)
2016-10-17 10:53:50 scdaemon[2939] DBG: Removal of a card: 0
2016-10-17 10:53:50 scdaemon[2939] detected reader 'Yubico Yubikey 4 OTP+U2F+CCID'
2016-10-17 10:53:50 scdaemon[2939] pcsc_connect failed: sharing violation (0x8010000b)
2016-10-17 10:53:50 scdaemon[2939] DBG: Removal of a card: 0
2016-10-17 10:53:50 scdaemon[2939] detected reader 'Yubico Yubikey 4 OTP+U2F+CCID'
2016-10-17 10:53:50 scdaemon[2939] pcsc_connect failed: sharing violation (0x8010000b)
2016-10-17 10:53:50 scdaemon[2939] DBG: Removal of a card: 0
2016-10-17 10:53:50 scdaemon[2939] new_reader_slot: out of slots
2016-10-17 10:53:50 scdaemon[2939] new_reader_slot: out of slots
2016-10-17 10:53:50 scdaemon[2939] new_reader_slot: out of slots
[...SNIP...]
[...SAME MESSAGE REPEATS...]
[...SNIP...]
[...SCDAEMON KILLED HERE...]
2016-10-17 11:23:23 scdaemon[2939] SIGTERM received - shutting down ...
2016-10-17 11:23:23 scdaemon[2939] scdaemon (GnuPG) 2.1.15 stopped
[...RE-RUN `gpg2 --card-status` HERE...]
2016-10-17 11:23:23 scdaemon[2939] scdaemon (GnuPG) 2.1.15 stopped
2016-10-17 11:23:28 scdaemon[19552] detected reader 'Yubico Yubikey 4 OTP+U2F+CCID'
2016-10-17 11:23:28 scdaemon[19552] pcsc_control failed: not transacted (0x80100016)
2016-10-17 11:23:28 scdaemon[19552] pcsc_vendor_specific_init: GET_FEATURE_REQUEST failed: 65547
2016-10-17 11:23:28 scdaemon[19552] updating reader 0 (0) status: 0x0000->0x0007 (0->1)
2016-10-17 11:23:28 scdaemon[19552] sending signal 31 to client 527
[...SCDAEMON KILLED HERE...]
2016-10-17 11:23:52 scdaemon[19552] SIGTERM received - shutting down ...
2016-10-17 11:23:52 scdaemon[19552] scdaemon (GnuPG) 2.1.15 stopped

[...RE-RAN `notary init -D` HERE...]
[...IT WORKED!...]
```

There is an [unaccepted patch which some say alleviates this issue][4].  However, it was not accepted by official upstream GnuPG, and the bug was closed as `wontfix`.  The reasoning was as follows:

> The patch is a work for problem somewhere in the PC/SC implementaion.  I am also
> not sure whether a pthread_cancel for a buggy PC/SC library is a good idea. 
> Terminating the process seems to be a better solution.
> 
> If gpgtools wants to apply this patch, they might of course do so but I don't
> want to apply it upstream in particular not to an older version (2.1 is current). 

There are [reported to be other workarounds][5] which require disabling OSX built-in `ifdhandler` (`sudo launchctl unload /System/Library/LaunchDaemons/com.apple.ifdreader.plist`).  GPGTools recommended to this user to try the latest nightly build as of Jan 26th 2015, and it was accepted that it was fixed.

However, because I'm using `gpg21` from Homebrew for `gpg-agent` socket forwarding support (`brew tap homebrew/versions && brew install gnupg21`), I cannot benefit from the fixes in GPGTools.   GPGTools has stated that [they are considering switch to `gnupg 2.1`, but not in the near future][6].

# TLDR;

There are bugs with Yubikey OpenPGP access in upstream GPG `2.0.x` AND `2.1.x` series releases on at least OSX versions:

- `10.10` (Yosemite)
- `10.11` (El Capitan)
- `10.12` (Sierra) [GPGTools not compatible with this release yet][7]!

`scdaemon` is launched to access the Yubikey smart card, and then begins to fail to access it after a period of time.  `scdaemon` also locks up access to the card while it is running by other applications including at least:

- Yubico Authenticator
- `notary`
- Possibly other tools needing access to Yubikey PIV slots?

GNUPG accepted workaround: `kill $(pidof scdaemon)`

[1]: https://gpgtools.tenderapp.com/discussions/problems/30646-gpg-agent-gets-stuck-when-used-with-smartcards-in-ssh-agent-mode
[2]: https://gpgtools.tenderapp.com/discussions/problems/28634-gpg-agent-stops-working-after-osx-upgrade-to-yosemite
[3]: https://bugs.gnupg.org/gnupg/issue1969
[4]: https://github.com/Homebrew/legacy-homebrew/pull/39248
[5]: https://gpgtools.tenderapp.com/discussions/problems/28634-gpg-agent-stops-working-after-osx-upgrade-to-yosemite#comment_35808149
[6]: https://gpgtools.tenderapp.com/discussions/feedback/1965-gpg-21
[7]: https://gpgtools.tenderapp.com/kb/gpgmail-faq/gpgmail-no-longer-working-after-macos-update

**Comment by endophage** [comment]
Awesome detective work! We haven't really tried out using our Yubikeys for GPG so we haven't run into this before. @a-dma I linked above is a member of the Yubico engineering team and if he doesn't know about it already (which would surprise me) will likely be very interested in the above info.

**Comment by trinitronx** [comment]
@endophage:  Happy to provide the info & observations!  As it seems now (at least on OSX), the current climate of GPG + Yubikey + other PIV tools is such that it is very difficult to get a seamless experience when using the Yubikey with all of these applications.  This is especially true when combined with the functionality of modern `gnupg 2.1` for the new `gpg-agent` socket forwarding over SSH feature.  So this makes it very difficult to sign code on a remote machine by forwarding the Yubikey -> `scdaemon` -> `gpg-agent` over SSH.  It also means that other applications that want direct access to the hardware must constantly fight with `scdaemon` to get the access.

It seems like a case of many competing implementations of libraries (PS/SC, CCID, `libykcs11`, etc...) are all fighting for access to the hardware, and do not play well with each other from a user experience perspective.

Hope it helps to track things down & smooth things out in the future!

**Comment by a-dma** [comment]
Hi, no worries, glad to help if I can.

Unfortunately what @trinitonx says is correct. The culprit is `gpg-agent` accessing smart cards in an exclusive way. According to the GnuPG folks, this is a design choice and a security feature, so probably not likely to change anytime soon. The agent implementation in `v2.1` is leaps and bounds ahead of what it used to be, but still there is no easy way of dealing with device access. The suggested solution is killing the agent and have it restart automatically by gpg whenever it is needed. This however does not include ssh since it doesn't have any knowledge of gpg-agent. In that case the agent should be restarted manually with something like `gpg --card-status` or `gpg-connect-agent /bye`.

**Comment by endophage** [comment]
I've created an issue specifically for documenting the gpg-agent exclusivity causing problems. Closing this.

**Comment by noeljackson** [comment]
For whomever might find this of help. I had the same issue on Debian. But, using pre-built notary binaries did and killing scdaemon first did not help. Building notary from scratch did work, and did require killing scdaemon and then initializing.

**Comment by trinitronx** [comment]
Yeah, it seems that the culprit is `scdaemon` which gets started by `gpg-agent` dynamically when some client to `gpg-agent` needs to perform some operation which needs access to the yubikey.  In my experience, killing `scdaemon` has always been a reliable workaround, until I perform some other operation that needs access to the GPG keys on the Yubikey.  The Yubico Authenticator app on Mac also seems to be smart enough to know this trick.  There is a setting to always kill `scdaemon` when launching this app.  I think this points to the root cause being `scdaemon` locking exclusive access to the yubikey, which prevents other applications from accessing any of the other slots or smart card operations.
